File: csharp/Database/ApplicationDatabase/Models/Role.cs

namespace Database.ApplicationDatabase.Models;

public enum RoleEnum 
{
    ADMIN,
    MODERATOR,
    NPC,
    PLAYER,
    GUEST
}

public class Role
{
    public required RoleEnum Name;

    public static Role Player() => new(){Name=RoleEnum.PLAYER};
    public static Role Admin() => new(){Name=RoleEnum.ADMIN};
    public static Role Npc() => new(){Name=RoleEnum.NPC};
    public static Role Moderator() => new(){Name=RoleEnum.MODERATOR};
    public static Role Guest() => new(){Name=RoleEnum.GUEST};

    public Role[] AboveIncluding() => Name switch
    {
        RoleEnum.ADMIN      => [Role.Admin()],
        RoleEnum.MODERATOR  => [Role.Moderator(), Role.Admin()],
        RoleEnum.NPC        => [Role.Npc(), Role.Moderator(), Role.Admin()],
        RoleEnum.PLAYER     => [Role.Player(), Role.Npc(), Role.Moderator(), Role.Admin()],
        RoleEnum.GUEST      => [Role.Guest(), Role.Player(), Role.Npc(), Role.Moderator(), Role.Admin()],
        _                   => [],
    };
}

File: backend/lib/auth_server/dal/models/Role.go
package models

import (
	"errors"
	"strings"
)

type Role int

const (
	ADMIN Role = iota
	MODERATOR
	NPC
	PLAYER
	GUEST
)

func (d Role) String() string {
	return [...]string{"admin", "moderator", "npc", "player", "guest"}[d]
}

func FromString(roleStr string) (Role, error) {
	switch strings.ToLower(roleStr) {
	case "admin":
		return ADMIN, nil
	case "moderator":
		return MODERATOR, nil
	case "npc":
		return NPC, nil
	case "player":
		return PLAYER, nil
	case "guest":
		return GUEST, nil
	default:
		return 0, errors.New("invalid role: " + roleStr)
	}
}
